retrieval/discovery/serverset.go

```diff
@@ -279,16 +280,20 @@ func (tc *zookeeperTreeCache) loop(failureMode bool) {
 				if err != nil {
 					log.Errorf("Error during processing of Zookeeper event: %s", err)
 					failure()
+				} else if tc.head.data == nil {
+					log.Errorf("Error during processing of Zookeeper event: path %s no longer exists", tc.prefix)
+					failure()
 				}
 			}
 		case <-retryChan:
 			log.Infof("Attempting to resync state with Zookeeper")
 			err := tc.recursiveNodeUpdate(tc.prefix, tc.head)
-			if err == nil {
-				failureMode = false
-			} else {
+			if err != nil {
 				log.Errorf("Error during Zookeeper resync: %s", err)
 				failure()
+			} else {
+				log.Infof("Zookeeper resync successful")
```

Comment:
I added it as it wasn't at all obvious from the logs if a resync had worked, or stalled somehow.